php/libraries/User.class.inc

```diff
@@ -165,10 +165,19 @@ class User extends UserPermissions implements
      */
     public static function &singleton(?string $username = null): \User
     {
-        static $instance;
-        if (is_null($instance)) {
-            $instance = User::factory($username);
+        static $currentuser = null;
+        static $allusers    = [];
+        if ($username === null && $currentuser !== null) {
+            return $currentuser;
         }
+        if ($username !== null && isset($allusers[$username])) {
+            return $allusers[$username ?? ''];
+        }
+        $instance = User::factory($username);
+        if ($currentuser === null) {
+            $currentuser = $instance;
+        }
```

Comment:
@driusan is this always correct? basically the first call to singleton will always be for logged in user and not some other username?

why not just maintain the array and not care about current user?